dot/core/service.go

```diff
@@ -548,3 +574,41 @@ func (s *Service) GetMetadata(bhash *common.Hash) ([]byte, error) {
 	rt.SetContextStorage(ts)
 	return rt.Metadata()
 }
+
+// RegisterRuntimeUpdatedChannel function to register chan that is notified when runtime version changes
+func (s *Service) RegisterRuntimeUpdatedChannel(ch chan<- runtime.Version) (byte, error) {
+	s.runtimeUpdateSubscriptionsLock.Lock()
+	defer s.runtimeUpdateSubscriptionsLock.Unlock()
+
+	if len(s.runtimeUpdateSubscriptions) == 256 {
+		return 0, errors.New("channel limit reached")
+	}
+
+	var id byte
+	for {
+		id = generateID()
+		if s.runtimeUpdateSubscriptions[id] == nil {
+			break
+		}
+	}
+
+	s.runtimeUpdateSubscriptions[id] = ch
+	return id, nil
+}
+
+// UnregisterRuntimeUpdatedChannel function to unregister runtime updated channel
+func (s *Service) UnregisterRuntimeUpdatedChannel(id byte) bool {
+	ch, ok := s.runtimeUpdateSubscriptions[id]
```

Comment:
locking here will solve your concurrent read/write error and consequently avoid any case where the channel is closed.

Reply:
Yes, updated and it does fix that error.